### ISSUE TITLE
Write content: Chapter 20. HTTP/2

### ISSUE THREAD

**Issue by rviscomi** [HTTPArchive/almanac.httparchive.org#173]
Part | Chapter | Authors | Reviewers | Analysts | Metrics | Analysis
-- | -- | -- | -- | -- | -- | --
IV. Content Distribution | 20. HTTP/2 | @bazzadp | @bagder @rmarx @dotjs | @paulcalvano | #22 | #101

### Due by: October 7

- [X] Review the results spreadsheet (TBD) and work with your analyst(s) (@paulcalvano) if you have any questions about the data.
- [X] Write the contents of your chapter and get it peer reviewed (@bagder @rmarx @dotjs).
- [X] File a PR to merge your written contents in markdown format to the [src/content/en/2019](https://github.com/HTTPArchive/almanac.httparchive.org/tree/master/src/content/en/2019) directory.

Refer to the [Authors' Guide](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Authors'-Guide) for writing advice.

**Comment by rviscomi** [comment]
Hey @bazzadp I know @paulcalvano is still working on https://github.com/HTTPArchive/almanac.httparchive.org/pull/176. I understand if you won't be able to get this finished by the October 7 deadline. Hopefully we get that PR merged today to unblock you.

**Comment by bazzadp** [comment]
Already made a start @rviscomi . Hope to make some good progress this weekend and ideally have a first draft completed but we’ll see.

**Comment by rviscomi** [comment]
Perfect, thanks for being on top of this!

**Comment by bazzadp** [comment]
OK @bagder, @rmarx , @dotjs (and CC to @paulcalvano  and @rviscomi in case you want a read), I have a very initial draft here: https://github.com/bazzadp/almanac.httparchive.org/blob/http2_chapter/src/content/en/2019/http2.md

This is based on data in this spreadsheet: https://docs.google.com/spreadsheets/d/1z1gdS3YVpe8J9K3g2UdrtdSPhRywVQRBz5kgBeqCnbw/edit?usp=sharing (though I think the Alt Svc data needs an update @paulcalvano as had to rerun it myself as looked wrong).

I'd appreciate any feedback on, particularly on:
- technical errors
- stats I've misinterpreted
- anything interesting you think I've missed out on reporting on
- anything that's unclear
- text or stats I should cut (I always write too much and could do with some reining in!). Did I give too much explanation at the start? Conscious that we may get readers to the Almanac that have only a vague idea of what HTTP/2 is (if any!).
- better ways of presenting the data - there's a lot of tables in there, which is maybe a bit boring.
- typos - especially cause I've tried to write in American English even though it leaves a bad taste in my mouth 😉
- anything else that comes to mind!

Promise not to take offence so more feedback the better. I can always ignore it if I strongly disagree but would rather hear it than not so don't be shy.

Will mull it over myself for a bit and then submit a PR once I'm a bit more comfortable with it, which might make it easier to comment on specific bits of the text, but for now just add comments to this issue with any feedback. Or ping me on Twitter (@tunetheweb) if you want to talk something over.

**Comment by bagder** [comment]
I need to off with saying that I think this is an impressive chapter and I don't have a much to add. Hats off to @bazzadp for a job well done! I've read it from top to bottom several times and I've found no technical issue and the statistics and details seem to match with the text.

I have some very minor language nits or discussion points:

1. the phrase "performance boast" is used in several places. Isn't it supposed to be "performance boost" ? I'm not a native speaker, I might be wrong.
2. is it "a HTTP..." or "an HTTP...". The text uses 'a' and 'an' a bit interchangeably. It struck me that this entire work probably already has to decide on how to go with this since this will hardly be the only section that uses these words. As a non-native speaker I would go with 'an' but I know that the 'hage' way of pronouncing the letter 'H' is also common...
3. (in the "What is HTTP/2?" section) "One reasons for this stagnation". An 's' too much.

In summary: a :+1: from me!

**Comment by bazzadp** [comment]
Thanks @bagder. You are correct in all your nits and will fix them.

IMHO it should be "an HTTP" - one of the quirks of English when dealing with a silent H, as I pronounce it Ach Tee Tee Pee rather Hach Tee Tee Pee. The [style guide we were asked to follow agrees with an example of an HTML](https://developers.google.com/style/articles)

**Comment by rmarx** [comment]
Hey Barry,

Great job on the post! Some very nice analyses of the different results and some bold statements (poor Apache!). It's definitely a good summary of the state of HTTP/2 in  2019. Thank you for the work!

I do have a couple of things I don't agree with fully (some more important than others) or which could use additional text, which I've listed below. Additionally, I have taken the liberty of doing a full editorial review of the post in a separate PR (see https://github.com/bazzadp/almanac.httparchive.org/pull/1). This shouldn't change any of the content/meaning, but just the language. For all remarks: feel free to do with them as you please :)

___

Note sure I agree with this statement entirely (and you don't really introduce it as such earlier):
>  The proprietary format (SPDY) was binary-based rather than text-based which solved some of the main performance problems with HTTP/1.1

I would rather highlight the fact it uses 1 vs 6 TCP connections, which I don't seem to find stated clearly.
___

This could use a reference:
> an estimated 95% of global users support HTTP/2 now
___

An outward link might be nice here:
> soon to be standardized as HTTP/3
___

The section that starts with "HTTP/2 is only supported by browsers over HTTPS" is conceptually correct, but I'm not sure you should talk about HTTPS instead of TLS (though kind of depends on target audience). This happens towards the end as well, as you talk about "negotiating HTTP/2 during the HTTPS handshake". It's probably fine if you keep HTTPS though.

___

In the "Cloudflare is the most popular CDN" part, you might mention that Cloudflare also uses (a modified version of) nginx under the hood (e.g., see https://blog.cloudflare.com/nginx-structural-enhancements-for-http-2-performance/)

___

I don't understand the logic behind the "number of requests has dropped, this is due to HTTP/2" argument... if HTTP/2 removes the need for bundling, you would see request count going up, right? You provide the right reasons for why this is (probably) not happening, but so I find the conclusion a bit weirdly worded. Why would H2 lead to request counts going down? 
___

I don't fully agree with this statement:
> we see around 100 sites pushing video - which I would say is definitely not the intended use case of HTTP/2 push

There are papers that have shown that pushing short initial segments of a video can dramatically reduce playback start time (e.g., https://biblio.ugent.be/publication/8552455/file/8552457.pdf)

___

I'm not quite sure what you mean with this sentence:
> Tying software to operating systems always lengthens deployment time - and in fact one of the very reasons for QUIC is to break this barrier with deploying TCP changes

There were no TCP changes needed for HTTP/2, neither are HTTP/2 implementations tied to the OS afaik. It's just that distributions don't bundle the latest versions of the implementations, something that I feel will not improve with QUIC... if anything, it will be worse with QUIC, seeing the large amount of changes are needed to the SSL libraries (e.g., OpenSSL) that are also tightly coupled to OS distributions (though, H2 had similar issues with ALPN support, so maybe still similar).

**Comment by bazzadp** [comment]
> Great job on the post! Some very nice analyses of the different results and some bold statements (poor Apache!). It's definitely a good summary of the state of HTTP/2 in 2019. Thank you for the work!

Thanks very much! I did try to balance out the Apache slating with a push for it later, but it is frustrating. And I like Apache and use it as my primary web server at work and in play!

> Additionally, I have taken the liberty of doing a full editorial review of the post in a separate PR (see bazzadp#1). This shouldn't change any of the content/meaning, but just the language.

Cool. Thanks for the extra effort in submitting as a PR. I may end up rejecting as I have another working copy (which includes some of the changes you pointed out!) so don't be offended by that if I do, and normally would have totally appreciated this extra effort.
 
> Note sure I agree with this statement entirely (and you don't really introduce it as such earlier):
> 
> > The proprietary format (SPDY) was binary-based rather than text-based which solved some of the main performance problems with HTTP/1.1
> 
> I would rather highlight the fact it uses 1 vs 6 TCP connections, which I don't seem to find stated clearly.

I meant the binary format allows multiplexing. Will have a look if I can make that clearer.

> The section that starts with "HTTP/2 is only supported by browsers over HTTPS" is conceptually correct, but I'm not sure you should talk about HTTPS instead of TLS (though kind of depends on target audience). This happens towards the end as well, as you talk about "negotiating HTTP/2 during the HTTPS handshake". It's probably fine if you keep HTTPS though.

Yeah, I've talked to you before about the fact I before to use the term HTTPS rather than SSL or TLS or SSL/TLS unless specifically talking about TLS-specific items. Mainly cause TLS is (still!) not widespread as many still use SSL and I can't be bother getting into that debate! It's a convention I have that may not be liked by all. I think it's totally relevant for this first sentence but I agree it could be argued that I should call it the TLS-handshake later but one of the reasons I don't do that is to avoid introducing another term (though I imagine most readers will be fine with TLS). Will think on it but probably just leave it as HTTPS. Auther's prerogative ! ;-)

> In the "Cloudflare is the most popular CDN" part, you might mention that Cloudflare also uses (a modified version of) nginx under the hood (e.g., see https://blog.cloudflare.com/nginx-structural-enhancements-for-http-2-performance/)

Potentially. I also feel it's moved quite far away from it and is a lot better (as that post shows!) so not sure it's really that beneficial to say it's like nginx anymore. Maybe @dotjs can comment on this?

> I don't understand the logic behind the "number of requests has dropped, this is due to HTTP/2" argument... if HTTP/2 removes the need for bundling, you would see request count going up, right? You provide the right reasons for why this is (probably) not happening, but so I find the conclusion a bit weirdly worded. Why would H2 lead to request counts going down?

Hmmmm.... not sure what I meant with this to be honest. Think I might have been tired and stat and word blind by this point as you are correct. In theory HTTP/2 should allow less bundling so more requests. This one is a definite error so will fix.

> I don't fully agree with this statement:
> 
> > we see around 100 sites pushing video - which I would say is definitely not the intended use case of HTTP/2 push
> 
> There are papers that have shown that pushing short initial segments of a video can dramatically reduce playback start time (e.g., https://biblio.ugent.be/publication/8552455/file/8552457.pdf)

Interesting. Will have a read and consider rewording. I'm trying to avoid citing deep research papers though as feel this should be at a higher level.

> I'm not quite sure what you mean with this sentence:
> 
> > Tying software to operating systems always lengthens deployment time - and in fact one of the very reasons for QUIC is to break this barrier with deploying TCP changes
> 
> There were no TCP changes needed for HTTP/2, neither are HTTP/2 implementations tied to the OS afaik. It's just that distributions don't bundle the latest versions of the implementations

I agree! But the fact is it is not easy to install latest Apache because it IS effectively tied to the distro as most don't want to install from source or from third party repos and, as I say on the next sentence:

"In many instances there is no real reason to tie server versions to operating systems. Apache (to use one of the more popular examples) will run with HTTP/2 support in older operating systems but getting an up to date version on to the server should not require the expertise or risk it currently does."

> something that I feel will not improve with QUIC... if anything, it will be worse with QUIC, seeing the large amount of changes are needed to the SSL libraries (e.g., OpenSSL) that are also tightly coupled to OS distributions (though, H2 had similar issues with ALPN support, so maybe still similar).

Yes maybe should reconsider this. I meant that TCP changes ARE ingrained to O/S and probably really NEED to be since they are integral to so much, so it's even worse than web servers (which ARE often tied to O/S but don't NEED to be). QUIC aims to solve the TCP stagnation issue by allowing faster iteration - once it's on the platform. Which obviously has an immeadiate barrier to getting it on the platform for the reasons you state. Will consider whether to reword or remove this to avoid confusing these similar but different issues.

Thanks again for the in depth review! I've now heard back from all three reviewers (@bagder , @dotjs  (though travelling at the mo so not had a chance to give full comments yet) and @rmarx ) so will make the changes and submit the PR. Feel free to comment further on that, even it's only a LGTM - in fact especially if it's that ;-)

**Comment by colinbendell** [comment]
I ran into the same issue with the protocol being empty. It turns out there are a number of bugs with the way that protocol version is reported in WPT, so I had to add a number of checks. Sometimes it was a HTTP2 socket, but for some reason WPT would report the http version as HTTP/1. Or more often, if the socket was a HTTP/1 TLS socket that used TLS resumption, then the protocol version would be empty. Take a look at my updated queries and results in the cdn chapter for better data output.

**Comment by bazzadp** [comment]
@colinbendell I found a number of reasons - service workers calls (reminds me of [my answer here](https://stackoverflow.com/questions/54513933/requests-served-from-serviceworker-are-downgraded-from-http-2-to-http-1-1)), cancelled requests, and lots I couldn't explain. Seemed to me mostly for Desktop - Mobile was cleaner.
